Allow the host_os.py script to execute in parallel

olavphibm opened the pull request:
The only restriction is that two processes cannot build the same package at the same time.

Commit:
Create a variable for common Mock arguments

This avoids repeating the same arguments at every command call.
Also, rename some variables to keep consistency and their meanings
clearer.

Commit:
Add unique extension to Mock's chroots

This allows multiples processes to use Mock concurrently, by creating a
separate chroot for each.

Commit:
Make repository cloning/checkout lazy

This is to prepare for concurrent build script executions.

Since the source code repository is a shared resource among builds on
the same system, if the repositories were all cloned and checked out
before any package was built, we wouldn't gain much by executing build
scripts concurrently because the second process would have to wait for
the first to finish using the repository (archiving its contents) to be
able to modify the repositories directories (e.g. checking out a
different commit).
By cloning/checking out only when the package will start being built, we
reduce the time that the shared repository is in use by a given process.

Commit:
Use lock files to allow concurrent build script executions

Using lock files, we are able to mark when the packages' shared resources
are being manipulated (the repository), so a second process can safely
execute operations on other packages and stall execution when it reaches
a locked package.
Since the whole package build is inside the critical section, this
creates a pipeline where only one process deals with a package at a
time.

mopsfelder commented on lib/mockbuilder.py:
```diff
@@ -41,25 +41,28 @@ def __init__(self, config):
+            "%(binary)s -r %(config_file)s %(extra_args)s "
+            "--uniqueext %(suffix)s" % dict(
+                binary=MOCK_BIN, config_file=self.mock_config,
+                extra_args=self.mock_args, suffix=self.timestamp))
```
What if we added these `common_mock_args` in `config.yaml` file under `mock_args` parameter?

olavphibm replied:
I don't think we should merge all the arguments together. The binary path and config file are required and so I think each one of them deserves a separate parameter. I'm adding those to the config file. The mock_args are optional, so I'm leaving as is. The timestamp is generated dynamically, so a parameter for it does not make much sense.

mopsfelder commented on lib/package.py:
```diff
@@ -180,3 +184,31 @@ def _download_build_files(self):
+        try:
+            fcntl.lockf(self.lock_file, fcntl.LOCK_EX | fcntl.LOCK_NB)
+        except IOError as exc:
+            if exc.errno == 11:
```
Where this number `11` comes from?  What if we declared a constant with a meaningful name to compare against `exc.errno`?

mopsfelder commented on lib/package.py:
```diff
@@ -180,3 +184,31 @@ def _download_build_files(self):
+        LOG.debug("Checking for lock on file {}.".format(self.lock_file_path))
+        self.lock_file = open(self.lock_file_path, "w")
+        try:
+            fcntl.lockf(self.lock_file, fcntl.LOCK_EX | fcntl.LOCK_NB)
```
What if `lock_file` is locked, i.e. opened by another process?  Shall we use any busy-waiting technique to wait until it can get the lock?

alanoe replied:
IMHO locking the source code repository when someone is using it is not an optimal solution. This only allows the simultaneous build of one package at once and only parallelizes builds of different packages. I'd prefer to use a  solution which parallelizes builds of *any* package , like using a dictionary (repo_name, commit_id) -> repo_directory_path

Commit:
Make Mock configuration file configurable

The configuration file was hardcoded in the CentOS class. This makes it
configurable by creating an attribute in the config.yaml file, similar
to the one used by the MockPungiSpinner class when creating an ISO.

olavphibm commented:
start build

alanoe commented on lib/mockbuilder.py:
```diff
@@ -33,33 +33,34 @@
 
 class Mock(build_system.PackageBuilder):
-    def __init__(self, config):
+    def __init__(self, mock_config):
```
This has nothing to do with the patch scope. Mention in longer commit msg that you are renaming variables for consistency. Also, minor point: mock_config_file or mock_config_file_path would be more intuitive IMHO

alanoe commented on lib/mockbuilder.py:
```diff
@@ -33,33 +33,34 @@
         self.archive = None
         self.timestamp = datetime.datetime.now().isoformat()
+        self.common_mock_args = (
+            "%(binary)s -r %(config_file)s %(extra_args)s" % dict(
```
Can't the extra_args be defined in the config file so that just need to call mock binary with the config file as parameter?

alanoe replied:
any reason to have different name for parameter name and parameter value (eg config_file and mock_config, extra_args and mock_args)? This makes code a little harder to understand

alanoe commented on config.yaml:
```diff
@@ -9,6 +9,7 @@ default:
  arch_and_endianness: 'ppc64le'
  release_notes_repo_url: "https://github.com/open-power-host-os/open-power-host-os.github.io.git"
  release_notes_repo_branch: "master"
+ mock_binary: "/usr/bin/mock"
```
I'm not sure what is the benefit of this patch. Pls document it in commit longer msg so it is clearer.

alanoe replied:
Is this a requirement? Is there a concrete use case (have you ever had the need of use an alternate Mock binary or found a mock rpm package which installs it in a different path) ?

olavphibm replied:
Some time ago, we had issues with the mock binary requiring root because we called the command `mock` instead of `/usr/bin/mock`, which lead to a different binary in some machines.
I remember someone commenting that it would be nice to have this configurable, so I think its harmless and maybe useful for someone to have this flexibility.

alanoe commented on config.yaml:
```diff
@@ -10,6 +10,7 @@ default:
  release_notes_repo_url: "https://github.com/open-power-host-os/open-power-host-os.github.io.git"
  release_notes_repo_branch: "master"
  mock_binary: "/usr/bin/mock"
+ mock_config: "./extras/centOS/7/mock/epel-7-ppc64le.cfg"
```
I do not understand this. Before, we had the ability to define a different Mock config file for each distro version. It seems to me we are losing this feature with this patch. -1 until motivation is clearer.

alanoe replied:
As we talked personally, there should be one Mock config file should be defined per distro and distro version. I noticed that the previous code (incorrectly) always used the same Mock config file for any distro version, but at least it used one different Mock config file for each distro. This patch allows one Mock config for all builds, reducing build script flexibility. Pls use subattribs in the yaml for 'centOS' and '7' as we have in the packages metadata (yaml) file.

olavphibm replied:
Done.

alanoe commented on lib/build_manager.py:
```diff
@@ -33,7 +33,8 @@ def __init__(self, packages_names, distro):
     def __call__(self):
```
The explanation in longer commit msg is confusing. You are assuming we'll use the same directory for cloning a repository and checking out a commit for different builds? This does not need to be necessary. We can keep a table of repo_name+commit_id -> directory_path and clone to a different directory whenever commit ID changes.

olavphibm replied:
Yes, I'm assuming there will be only one shared repository for each package for all builds. This avoids having to clone potentially big repositories multiple times. The actual build can be executed in parallel, since after the archive is ready the repository is not used anymore. I'm adding this improvement in a separate commit.
I'm improving the commit message.

alanoe replied:
New commit msg is much better!

alanoe commented on lib/distro.py:
```diff
@@ -57,8 +57,11 @@ def build_packages(self, packages):
         """
```
I have not found out how to comment the commit msg, so I'm doing it here. Use "lock files" instead of "lockfiles" pls, better English

Commit:
Make Mock binary configurable

The user may want to use an alternate Mock binary or have it installed
in a different path.

Commit:
Remove the CentOS 7.2 directory symlink

This symlink was created in commit
659651a56fba4f9a494afce0bad94e812c6d5b64
to allow older versions to be built, but it turned out it was not
needed because the Mock configuration file path is hardcoded in the
CentOS class.

olavphibm commented:
start build

alanoe commented on lib/mockbuilder.py:
```diff
@@ -42,9 +42,10 @@ def __init__(self):
         self.archive = None
```
In the longer commit msg, I'd prefer "This allows multiples processes to use Mock concurrently, creating a separate chroot for each". This makes clearer that different processes do *not* share a chroot.

olavphibm replied:
Done.

alanoe commented on lib/distro.py:
```diff
@@ -57,8 +57,11 @@ def build_packages(self, packages):
         """
         self.package_builder.initialize()
         for package in packages:
+            package.lock()
```
Do we need to lock if the package uses only the url source?

olavphibm replied:
No. I've added a commit to disable locking for those packages.

alanoe commented on lib/distro.py:
```diff
@@ -57,10 +57,10 @@ def build_packages(self, packages):
             package.lock()
             package.download_files(recurse=False)
             self.package_builder.build(package)
-            package.unlock()
```
It is simpler to understand the locking mechanism if lock() and unlock() are in the same level. I suggest moving the source code packaging to outside package_builder.build().

olavphibm replied:
I agree with that, but I don't want to break the build function in too many steps and make the Mock class' usability harder. I've decided to externalize the `Mock._prepare()` function, with the small drawback that the build_files copying will be included in the lock.

Commit:
Anticipate the package's lock file unlock

After the repository is check out at the specified commit and archived
to a .tar.gz file, it is no longer used, so the package's lock can be
safely released.
To make it easier to spot the related lock and unlock calls and avoid
breaking the build function in too many public functions, I've
externalized the _prepare() function and renamed it to
prepare_sources(), allowing the lock to be released after the archive is
ready and the extra build files are copied to the build directory, which
should take a very small time compared to the actual build.
This allows the same package to be built by multiple processes in
parallel, while avoiding the need to clone the repository multiple
times.

Commit:
Disable lock files for packages that have exclusively URL sources

The sources are not downloaded to a shared location in this case.

alanoe commented on lib/mockbuilder.py:
```diff
@@ -57,7 +57,6 @@ def initialize(self):
 
     def build(self, package):
         LOG.info("%s: Starting build process" % package.name)
-        self._prepare(package)
```
why this line was removed in this patch? It seems to me it should have been removed in the previous patch

olavphibm replied:
That was a mistake indeed. The PR is already merged though, so there's nothing I can do.
